Fix Lambda Extension logs shutdown

nhinsch opened the pull request:
### What does this PR do?

There is a race condition that causes panics and prevents us from receiving logs when shutting down the Lambda Extension. When the extension receives a shutdown event from the API, it stops the logs agent. However, the extension may continue to receive log messages from the logs API after receiving the shutdown event. Currently, when these messages are received, we try to send them to the logs agent. Since the logs agent has been stopped, this results in a panic.

To fix this issue:
- Consolidate the shutdown logic in a new `Daemon.Stop()` function.
- Add a 1-second delay after we receive the shutdown signal before shutting down. This should allow for final log messages to be delivered from the logs API.
- After the 1-second delay, shut down the HTTP server first before shutting down the agents. This will prevent us from sending messages to the agents after they have already shut down.
- Do some refactoring so that we also call `Daemon.Stop()` after receiving a `SIGINT` signal. This shouldn't really happen in practice because during a shutdown, we receive a shutdown message over the Lambda API, and then `SIGKILL` two seconds later.

### Motivation

The current behavior causes panics and lost log messages.

### Describe how to test your changes

I have been manually testing these changes with functions running periodically in the sandbox account that frequently encounter the race condition. I used debug log messages to confirm the behavior is as expected.

nhinsch committed:
Add delay and suspension

DarcyRaynerDD commented on pkg/serverless/protocol.go:
```diff
@@ -244,6 +260,14 @@ func (l *LogsCollection) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 				logMessage := logConfig.NewChannelMessageFromLambda([]byte(message.StringRecord), message.Time, arn, lastRequestID, functionName)
+
+				// Do not publish logs to channel if logs collection has been suspended
+				if l.daemon.logsCollectionSuspended {
```
What happens if the logs channel is shutdown after logsCollectionSuspended is checked. I think that's another race condition. Would it be better to shutdown the Daemon HTTP API first, then shutdown the logs channel?

nhinsch replied:
Yes, I see your point here. I tried to minimize that by doing the check of `logsCollectionSuspended` at the last possible moment before we attempt to send the log message to the agent (rather than doing the check at the beginning of the API endpoint, when it would more likely be stale by the end). But I see that the race condition is not completely eliminated.

When you say shut down the Daemon HTTP API first, to clarify, I don't think we have a way to shut it down completely right now, are you proposing adding that? And then making sure that shutdown is completed before we begin the shutdown of the logs agent?

nhinsch replied:
Looks like the `http.Server` already has a `Shutdown()` method -- let me see what I can do!

DarcyRaynerDD commented on cmd/serverless/main.go:
```diff
@@ -481,15 +481,14 @@ func readAPIKeyFromSSM() (string, error) {
-func stopCallback(cancel context.CancelFunc) {
-	// gracefully shut down any component
+func shutDownGracefully(cancel context.CancelFunc) {
 	cancel()
```
Maybe we should leave the shutdown logic to the daemon, and have that control the shutdown order

nhinsch replied:
We could do that -- I think this is here because it is called with `defer` in the main `run()` function, so that it always executes, even when we receive a `SIGINT` from the environment. But since none of our other graceful shutdown logic would run in that case, maybe it's better to just consolidate it in the Daemon.

nhinsch committed:
Refactor

nhinsch committed:
Fix comment

nhinsch committed:
Fix pointers

nhinsch committed:
Handle error

nhinsch commented:
@DarcyRaynerDD I have refactored this and incorporated your suggestions. I updated the initial PR comment to reflect the new approach. Let me know your thoughts.

remeh commented on pkg/serverless/protocol.go:
```diff
@@ -139,19 +147,58 @@ func (d *Daemon) TriggerFlush(ctx context.Context, shutdown bool) {
+		log.Error("Error shutting down HTTP server")
+	}
+
+	err = aws.PersistCurrentStateToFile()
```
That's a detail but you may want to do that before the `sleep`, just in case the extension is being forced shutdown by the environment during the sleep.

nhinsch replied:
I think it is better this way, in order to handle the scenario where we receive logs during the 1-second window that change the state. In other words, I am trying to avoid the following situation:

1. We persist the state before sleeping
2. We receive logs during the state that change the current request ID, but we can't persist it
3. We shut down
4. Upon restarting after the timeout, we restore the incorrect state

remeh replied:
Got your point 👍

remeh commented on pkg/serverless/protocol.go:
```diff
@@ -139,19 +147,58 @@ func (d *Daemon) TriggerFlush(ctx context.Context, shutdown bool) {
+func (d *Daemon) Stop() {
+	// Can't shut down before starting
+	// If the DogStatsD daemon isn't ready, wait for it.
+	d.ReadyWg.Wait()
```
I think this is the correct thing to do, however, I'm slightly concerned that a shutdown right after a failed startup may lock here if the daemon is never ready. 

I've checked and I think there is a scenario where it could happen if the trace agent fails to start here: https://github.com/DataDog/datadog-agent/blob/master/cmd/serverless/main.go#L359
This return should probably be removed to avoid having an issue in the future. 

I just want to emphase that using this lock in the shutdown can be tricky / slightly risky, but still, probably the right thing to do.

nhinsch replied:
I went ahead and removed the return.

I see your point but also note that the Lambda Platform will kill the extension after two seconds during shutdown.

remeh reviewed (changes requested):
Just a comment on the usage of the lock in the shutdown which I think needs one more commit, but except from that, solid!

nhinsch committed:
Rm return

remeh reviewed (approved):
💯
